bpo-43497: Emit SyntaxWarnings for assertions with tuple constants.

tsukasa-au opened the pull request:
Add a test that shows that a tuple constant (a tuple, where all of its
members are also compile-time constants) produces a SyntaxWarning. Then
fix this failure.

<!-- issue-number: [bpo-43497](https://bugs.python.org/issue43497) -->
https://bugs.python.org/issue43497
<!-- /issue-number -->

tsukasa-au committed:
bpo-43497: Emit SyntaxWarnings for assertions with tuple constants.

Add a test that shows that a tuple constant (a tuple, where all of its
members are also compile-time constants) produces a SyntaxWarning. Then
fix this failure.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

markshannon commented on Lib/test/test_grammar.py:
```diff
@@ -1211,6 +1211,8 @@ def testAssert2(self):
         self.check_syntax_warning('assert(x, "msg")',
                                   'assertion is always true')
+        self.check_syntax_warning('assert(False, "msg")',
+                                  'assertion is always true')
```
Could you also add to test that this doesn't warn on `assert False, "msg"`.

These tests shouldn't be guarded by  `@unittest.skipUnless(__debug__, ...)`.
The warning should be issued regardless of `__debug__`.
See below.

tsukasa-au replied:
Done.

markshannon commented on Python/compile.c:
```diff
@@ -3357,8 +3357,11 @@ compiler_assert(struct compiler *c, stmt_ty s)
+        asdl_seq_LEN(s->v.Assert.test->v.Tuple.elts) > 0) ||
+        (s->v.Assert.test->kind == Constant_kind &&
+         PyTuple_Check(s->v.Assert.test->v.Constant.value) &&
+         PyTuple_Size(s->v.Assert.test->v.Constant.value) > 0)) 
```
By moving the `if (c->c_optimize)` check after these checks, we can warn even when `__debug__ == False`

tsukasa-au replied:
Done.

tsukasa-au committed:
Make SyntaxWarnings also work when "optimized".

Also add some additional unit tests while we are here.

tsukasa-au committed:
Split tests for SyntaxWarning to SyntaxError conversion

SyntaxWarnings emitted by the compiler when configured to be errors are
actually raised as SyntaxError exceptions.

Move these tests into their own method and add a test to ensure they are
raised. Previously we only tested that they were not raised for a
"valid" assertion statement.

tsukasa-au committed:
Remove trailing whitespace accidentally added.

The first commit in this patch accidentally added a trailing whitespace.
Remove it.
